obs-mcp/pkg/prometheus/guardrails.go

```diff
@@ -102,27 +119,155 @@ func (g *Guardrails) IsSafeQuery(query string) (bool, error) {
 				// Rule 3: Check for expensive regex matchers on *any* label i.e blanket matchers
 				if g.DisallowBlanketRegex {
 					isRegex := m.Type == labels.MatchRegexp || m.Type == labels.MatchNotRegexp
-					if isRegex {
-						if m.Value == ".*" || m.Value == ".+" {
-							foundUnsafe = true
-							return fmt.Errorf("unsafe")
+					if isRegex && (m.Value == ".*" || m.Value == ".+") {
+						// If MaxLabelCardinality is set, defer the check to TSDB lookup
+						if g.MaxLabelCardinality > 0 {
+							hasBlanketRegex = true
+							blanketRegexLabels = append(blanketRegexLabels, m.Name)
+						} else {
+							// MaxLabelCardinality is 0, always disallow blanket regex here
+							unsafeReason = fmt.Errorf("query uses blanket regex (%s) on label %q, which is disallowed", m.Value, m.Name)
+							return unsafeReason
 						}
 					}
 				}
 			}
 
 			// Rule 2: All vector selectors must have at least one non-name label matcher
 			if g.RequireLabelMatcher && !hasNonNameMatcher {
-				foundUnsafe = true
-				return fmt.Errorf("unsafe")
+				metricName := n.Name
+				if metricName == "" {
+					metricName = "<metric>"
+				}
```

Comment:
IIUC this would be the case for `{...}` metric queries, where the metric name could be inferred from `__name__`'s value? `<metric>` doesn't seem very helpful as is.